dump error on ARRAY

```
cockroach dump test --certs-dir=/home/cockroach/certs/ --host 192.168.1.54 > test.sql
```
there is dump command.

```
create table test(
   values string array not null default array[]
);
insert into test(values) values(array['a','b','c']);
```
it works.


```
create table test(
   values int array not null default array[]
);
no data inserted.
```
it works.

```
create table test(
   values int array not null default array[]
);
insert into test(values) values(array[1,2,3,4,5]);
```
it didn't work.




error:

```
E171024 13:54:35.289938 1 util/log/crash_reporting.go:117  a panic has occurred!
panic: unknown []byte type: {}, values: INT[] [recovered]
        panic: unknown []byte type: {}, values: INT[]
goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.RecoverAndReportPanic(0x7f3a7b3d48d8, 0xc420014218, 0xc420432000)
        /go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:99 +0x9f
panic(0x1b44e00, 0xc4202f04e0)
        /usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/cli.dumpTableData(0x2ab4620, 0xc42000e018, 0xc420469a80, 0xc420271880, 0x1e, 0x174, 0xc4206cc580, 0x1, 0xc42026f330, 0x2, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/dump.go:478 +0x126d
github.com/cockroachdb/cockroach/pkg/cli.runDump(0x2a9d9e0, 0xc420428f80, 0x1, 0x4, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/dump.go:113 +0xa3a
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x2a9d9e0, 0xc420428f80, 0x1, 0x4, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:35 +0x5a
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).execute(0x2a9d9e0, 0xc420428e80, 0x4, 0x4, 0x2a9d9e0, 0xc420428e80)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:644 +0x3f8
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x2a9b820, 0x5d876e, 0xc42027f390, 0xc42072ff10)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:734 +0x339
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).Execute(0x2a9b820, 0xc42072ff00, 0xc42027f360)
        /go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:693 +0x2b
github.com/cockroachdb/cockroach/pkg/cli.Run(0xc4200100d0, 0x5, 0x5, 0xc420014218, 0xc420432000)
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:141 +0x6d
github.com/cockroachdb/cockroach/pkg/cli.Main()
        /go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:50 +0x18d
main.main()
        /go/src/github.com/cockroachdb/cockroach/main.go:26 +0x20
```

--- a/pkg/cli/dump.go
+++ b/pkg/cli/dump.go
@@ -433,7 +433,18 @@ func dumpTableData(w io.Writer, conn *sqlConn, clusterTS string, md tableMetadat
 					default:
 						// STRING and DECIMAL types can have optional length
 						// suffixes, so only examine the prefix of the type.
-						if strings.HasPrefix(md.columnTypes[cols[si]], "STRING") {
+						// In addition, we can only observe ARRAY types by their [] suffix.
+						if strings.HasSuffix(md.columnTypes[cols[si]], "[]") {
+							typ := strings.TrimRight(md.columnTypes[cols[si]], "[]")
+							elemType, err := parser.StringToColType(typ)
+							if err != nil {
+								return err
+							}
+							d, err = parser.ParseDArrayFromString(parser.NewTestingEvalContext(), string(t), elemType)
+							if err != nil {
+								return err
+							}
+						} else if strings.HasPrefix(md.columnTypes[cols[si]], "STRING") {
 							d = parser.NewDString(string(t))
 						} else if strings.HasPrefix(md.columnTypes[cols[si]], "DECIMAL") {
 							d, err = parser.ParseDDecimal(string(t))

--- a/pkg/cli/dump_test.go
+++ b/pkg/cli/dump_test.go
@@ -59,11 +59,12 @@ func TestDumpRow(t *testing.T) {
 		e decimal,
 		u uuid,
 		ip inet,
+		ary string[],
 		tz timestamptz,
 		e1 decimal(2),
 		e2 decimal(2, 1),
 		s1 string(1),
-		FAMILY "primary" (i, f, d, t, n, o, u, ip, tz, e1, e2, s1, rowid),
+		FAMILY "primary" (i, f, d, t, n, o, u, ip, ary, tz, e1, e2, s1, rowid),
 		FAMILY fam_1_s (s),
 		FAMILY fam_2_b (b),
 		FAMILY fam_3_e (e)
@@ -78,6 +79,7 @@ func TestDumpRow(t *testing.T) {
 		1.2345,
 		'e9716c74-2638-443d-90ed-ffde7bea7d1d',
 		'192.168.0.1',
+		ARRAY['hello','world'],
 		'2016-01-25 10:10:10',
 		3.4,
 		4.5,
@@ -113,22 +115,23 @@ CREATE TABLE t (
 	e DECIMAL NULL,
 	u UUID NULL,
 	ip INET NULL,
+	ary STRING[] NULL,
 	tz TIMESTAMP WITH TIME ZONE NULL,
 	e1 DECIMAL(2) NULL,
 	e2 DECIMAL(2,1) NULL,
 	s1 STRING(1) NULL,
-	FAMILY "primary" (i, f, d, t, n, o, u, ip, tz, e1, e2, s1, rowid),
+	FAMILY "primary" (i, f, d, t, n, o, u, ip, ary, tz, e1, e2, s1, rowid),
 	FAMILY fam_1_s (s),
 	FAMILY fam_2_b (b),
 	FAMILY fam_3_e (e)
 );
 
-INSERT INTO t (i, f, s, b, d, t, n, o, e, u, ip, tz, e1, e2, s1) VALUES
-	(1, 2.3, 'striiing', '\x613162326333', '2016-03-26', '2016-01-25 10:10:10+00:00', '2h30m30s', true, 1.2345, 'e9716c74-2638-443d-90ed-ffde7bea7d1d', '192.168.0.1', '2016-01-25 10:10:10+00:00', 3, 4.5, 's'),
-	(NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL),
-	(NULL, '+Inf', NULL, NULL, NULL, NULL, NULL, NULL, 'Infinity', NULL, NULL, NULL, NULL, NULL, NULL),
-	(NULL, '-Inf', NULL, NULL, NULL, NULL, NULL, NULL, '-Infinity', NULL, NULL, NULL, NULL, NULL, NULL),
-	(NULL, 'NaN', NULL, NULL, NULL, NULL, NULL, NULL, 'NaN', NULL, NULL, NULL, NULL, NULL, NULL);
+INSERT INTO t (i, f, s, b, d, t, n, o, e, u, ip, ary, tz, e1, e2, s1) VALUES
+	(1, 2.3, 'striiing', '\x613162326333', '2016-03-26', '2016-01-25 10:10:10+00:00', '2h30m30s', true, 1.2345, 'e9716c74-2638-443d-90ed-ffde7bea7d1d', '192.168.0.1', ARRAY['hello':::STRING,'world':::STRING], '2016-01-25 10:10:10+00:00', 3, 4.5, 's'),
+	(NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL),
+	(NULL, '+Inf', NULL, NULL, NULL, NULL, NULL, NULL, 'Infinity', NULL, NULL, NULL, NULL, NULL, NULL, NULL),
+	(NULL, '-Inf', NULL, NULL, NULL, NULL, NULL, NULL, '-Infinity', NULL, NULL, NULL, NULL, NULL, NULL, NULL),
+	(NULL, 'NaN', NULL, NULL, NULL, NULL, NULL, NULL, 'NaN', NULL, NULL, NULL, NULL, NULL, NULL, NULL);
 `
 
 	if out != expect {

--- a/pkg/sql/parser/parse_array.go
+++ b/pkg/sql/parser/parse_array.go
@@ -144,6 +144,41 @@ func (p *parseState) parseElement() error {
 	return p.result.Append(d)
 }
 
+// StringToColType returns a column type given a string representation of the
+// type. Used by dump.
+func StringToColType(s string) (ColumnType, error) {
+	switch s {
+	case "BOOL":
+		return boolColTypeBool, nil
+	case "INT":
+		return intColTypeInt, nil
+	case "FLOAT":
+		return floatColTypeFloat, nil
+	case "DECIMAL":
+		return decimalColTypeDecimal, nil
+	case "TIMESTAMP":
+		return timestampColTypeTimestamp, nil
+	case "TIMESTAMPTZ", "TIMESTAMP WITH TIME ZONE":
+		return timestampTzColTypeTimestampWithTZ, nil
+	case "INTERVAL":
+		return intervalColTypeInterval, nil
+	case "UUID":
+		return uuidColTypeUUID, nil
+	case "INET":
+		return ipnetColTypeINet, nil
+	case "DATE":
+		return dateColTypeDate, nil
+	case "STRING":
+		return stringColTypeString, nil
+	case "NAME":
+		return nameColTypeName, nil
+	case "BYTES":
+		return bytesColTypeBytes, nil
+	default:
+		return nil, pgerror.NewErrorf(pgerror.CodeInternalError, "unexpected column type %s", s)
+	}
+}
+
 // ParseDArrayFromString parses the string-form of constructing arrays, handling
 // cases such as `'{1,2,3}'::INT[]`.
 func ParseDArrayFromString(evalCtx *EvalContext, s string, t ColumnType) (*DArray, error) {

--- a/pkg/sql/sqlbase/structured.go
+++ b/pkg/sql/sqlbase/structured.go
@@ -1911,9 +1911,8 @@ func ColumnsSelectors(cols []ColumnDescriptor) parser.SelectExprs {
 	return exprs
 }
 
-// SQLString returns the SQL string corresponding to the type.
-func (c *ColumnType) SQLString() string {
-	switch c.SemanticType {
+func colTypeSQLString(c *ColumnType, semType ColumnType_SemanticType) string {
+	switch semType {
 	case ColumnType_INT:
 		if c.Width > 0 && c.VisibleType == ColumnType_BIT {
 			// A non-zero width indicates a bit array. The syntax "INT(N)"
@@ -1922,11 +1921,11 @@ func (c *ColumnType) SQLString() string {
 		}
 	case ColumnType_STRING:
 		if c.Width > 0 {
-			return fmt.Sprintf("%s(%d)", c.SemanticType.String(), c.Width)
+			return fmt.Sprintf("%s(%d)", semType.String(), c.Width)
 		}
 	case ColumnType_FLOAT:
 		if c.Precision > 0 {
-			return fmt.Sprintf("%s(%d)", c.SemanticType.String(), c.Precision)
+			return fmt.Sprintf("%s(%d)", semType.String(), c.Precision)
 		}
 		if c.VisibleType == ColumnType_DOUBLE_PRECISON {
 			return "DOUBLE PRECISION"
@@ -1934,9 +1933,9 @@ func (c *ColumnType) SQLString() string {
 	case ColumnType_DECIMAL:
 		if c.Precision > 0 {
 			if c.Width > 0 {
-				return fmt.Sprintf("%s(%d,%d)", c.SemanticType.String(), c.Precision, c.Width)
+				return fmt.Sprintf("%s(%d,%d)", semType.String(), c.Precision, c.Width)
 			}
-			return fmt.Sprintf("%s(%d)", c.SemanticType.String(), c.Precision)
+			return fmt.Sprintf("%s(%d)", semType.String(), c.Precision)
 		}
 	case ColumnType_TIMESTAMPTZ:
 		return "TIMESTAMP WITH TIME ZONE"
@@ -1949,12 +1948,17 @@ func (c *ColumnType) SQLString() string {
 		}
 		return fmt.Sprintf("%s COLLATE %s", ColumnType_STRING.String(), *c.Locale)
 	case ColumnType_ARRAY:
-		return c.ArrayContents.String() + "[]"
+		return colTypeSQLString(c, *c.ArrayContents) + "[]"
 	}
 	if c.VisibleType != ColumnType_NONE {
 		return c.VisibleType.String()
 	}
-	return c.SemanticType.String()
+	return semType.String()
+}
+
+// SQLString returns the SQL string corresponding to the type.
+func (c *ColumnType) SQLString() string {
+	return colTypeSQLString(c, c.SemanticType)
 }
 
 // MaxCharacterLength returns the declared maximum length of characters if the

--- a/pkg/sql/sqlbase/table.go
+++ b/pkg/sql/sqlbase/table.go
@@ -1780,11 +1780,7 @@ func parserTypeToEncodingType(t parser.Type) (encoding.Type, error) {
 		return encoding.Float, nil
 	case parser.TypeDecimal:
 		return encoding.Decimal, nil
-	case parser.TypeBytes:
-		return encoding.Bytes, nil
-	case parser.TypeString:
-		return encoding.Bytes, nil
-	case parser.TypeBytes:
+	case parser.TypeBytes, parser.TypeString, parser.TypeName:
 		return encoding.Bytes, nil
 	case parser.TypeTimestamp, parser.TypeTimestampTZ, parser.TypeDate:
 		return encoding.Time, nil

